From b52c4a0699b3801b96cd25567843faa6ab57daf3 Mon Sep 17 00:00:00 2001
From: Daniel Scharrer <daniel@constexpr.org>
Date: Mon, 13 May 2013 20:47:16 +0200
Subject: [PATCH] Fix crash on startup under Windows

Don't assume that the runtime initializes the __wargv global variable.
---
 src/util/windows.cpp | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/src/util/windows.cpp b/src/util/windows.cpp
index 7ee7caf..87cbaa9 100644
--- a/src/util/windows.cpp
+++ b/src/util/windows.cpp
@@ -27,6 +27,7 @@
 #include <clocale>
 
 #include <windows.h>
+#include <shellapi.h>
 
 #include <boost/filesystem/path.hpp>
 
@@ -52,9 +53,11 @@ int main() {
 	
 	std::setlocale(LC_ALL, "");
 	
-	// Get the UTF-16 command-line parameters and convert them to UTF-8 ourself.
-	int argc = __argc;
-	wchar_t ** wargv = __wargv;
+	// Emulate wmain() as it's nonstandard and not supported by MinGW.
+	int argc = 0;
+	wchar_t ** wargv = CommandLineToArgvW(GetCommandLineW(), &argc);
+	
+	// Convert the UTF-16 command-line parameters to UTF-8 ourself.
 	char ** argv = new char *[argc + 1];
 	argv[argc] = NULL;
 	for(int i = 0; i < argc; i++) {
@@ -63,6 +66,8 @@ int main() {
 		WideCharToMultiByte(CP_UTF8, 0, wargv[i], -1, argv[i], n, NULL, NULL);
 	}
 	
+	LocalFree(wargv);
+	
 	// Tell boost::filesystem to interpret our path strings as UTF-8.
 	std::locale global_locale = std::locale();
 	std::locale utf8_locale(global_locale, new utf8_codecvt);
-- 
1.9.1

